File: Node.js/lambda_calculus.js
/**
 * Lambda Calculus Implementation in Node.js
 * Created by Claude 3.5 Sonnet (Anthropic, PBC)
 * 
 * Recommended filename: lambda_calculus.js
 * 
 * Usage: 
 * 1. Save this file as lambda_calculus.js
 * 2. Run with Node.js: node lambda_calculus.js
 * 3. Alternatively, you can import functions from this file in your own Node.js scripts
 */

// Step 1: Understand Lambda Expressions
const identity = x => x;
console.log(identity(5)); // 5

// Step 2: Implement Church Booleans
const lcTrue = x => y => x;
const lcFalse = x => y => y;
const lcIf = condition => then => else_ => condition(then)(else_);

console.log(lcIf(lcTrue)('yes')('no')); // 'yes'

// Step 3: Implement Basic Combinators
const I = x => x; // Identity
const K = x => y => x; // Kestrel
const KI = x => y => y; // Kite

// Step 4: Create Church Numerals
const lcZero = f => x => x;
const lcSucc = n => f => x => f(n(f)(x));

const churchToInt = n => n(x => x + 1)(0);

console.log(churchToInt(lcSucc(lcZero))); // 1

// Step 5: Implement Basic Arithmetic
const lcAdd = m => n => f => x => m(f)(n(f)(x));
const lcMult = m => n => f => m(n(f));

const lcOne = lcSucc(lcZero);
const lcTwo = lcSucc(lcOne);

console.log(churchToInt(lcAdd(lcOne)(lcTwo))); // 3

// Step 6: Implement Pairs
const lcPair = x => y => f => f(x)(y);
const lcFirst = p => p(x => y => x);
const lcSecond = p => p(x => y => y);

const myPair = lcPair('hello')('world');
console.log(lcFirst(myPair)); // 'hello'

// Step 7: Build Lists Using Pairs
const lcNil = lcPair(lcTrue)(lcTrue);
const lcCons = x => xs => lcPair(lcFalse)(lcPair(x)(xs));
const lcIsNil = xs => lcFirst(xs);
const lcHead = xs => lcFirst(lcSecond(xs));
const lcTail = xs => lcSecond(lcSecond(xs));

// Step 8: Implement Comparison Operations
const lcIsZero = n => n(x => lcFalse)(lcTrue);
const lcLessThan = m => n => lcIsZero(lcSub(m)(n));
const lcEqual = m => n => lcAnd(lcLessThan(m)(n))(lcLessThan(n)(m));

// Helper functions for comparison operations
const lcAnd = p => q => p(q)(p);
const lcSub = m => n => n(lcPred)(m);
const lcPred = n => f => x => n(g => h => h(g(f)))(y => x)(I);

// Step 9: Implement Higher-Order List Operations
const lcMap = f => xs => lcIsNil(xs) ? lcNil : lcCons(f(lcHead(xs)))(lcMap(f)(lcTail(xs)));
const lcFilter = p => xs => lcIsNil(xs) ? lcNil : lcIf(p(lcHead(xs)))
                                                   (lcCons(lcHead(xs))(lcFilter(p)(lcTail(xs))))
                                                   (lcFilter(p)(lcTail(xs)));
const lcFold = f => acc => xs => lcIsNil(xs) ? acc : lcFold(f)(f(acc)(lcHead(xs)))(lcTail(xs));

// Step 10: Explore Recursion and Fixed-Point Combinators
const Y = f => (x => f(y => x(x)(y)))(x => f(y => x(x)(y)));

// Example: Factorial using Y combinator
const factorial = Y(f => n => lcIsZero(n) ? lcOne : lcMult(n)(f(lcPred(n))));

console.log(churchToInt(factorial(lcTwo))); // 2

// Helper function to convert JavaScript numbers to Church numerals
const intToChurch = n => {
    if (n === 0) return lcZero;
    return lcSucc(intToChurch(n - 1));
};

// Example usage of higher-order functions
const exampleList = lcCons(intToChurch(1))(lcCons(intToChurch(2))(lcCons(intToChurch(3))(lcNil)));
const doubleList = lcMap(n => lcMult(n)(lcTwo))(exampleList);
const evenList = lcFilter(n => lcIsZero(lcMod(n)(lcTwo)))(exampleList);
const sum = lcFold(lcAdd)(lcZero)(exampleList);

console.log('Double list:', churchToInt(lcHead(doubleList))); // 2
console.log('Even list:', churchToInt(lcHead(evenList))); // 2
console.log('Sum:', churchToInt(sum)); // 6

// Helper function for modulo operation
const lcMod = m => n => Y(f => a => b => lcLessThan(a)(b) ? a : f(lcSub(a)(b))(b))(m)(n);

File: Haskell/LambdaCalculus.hs
{-
Model: Claude 3.5 Sonnet (July 2024)
Author: Anthropic, PBC

Filename: LambdaCalculus.hs
Usage: Load this file into GHCi (Glasgow Haskell Compiler interactive environment) 
       by running `ghci LambdaCalculus.hs` in your terminal.
       Then you can interact with the defined functions and test the implementation.
-}

module LambdaCalculus where

-- Step 1: Understand Lambda Expressions
-- In Haskell, we can use anonymous functions for lambda expressions
identity :: a -> a
identity = \x -> x

-- Step 2: Implement Church Booleans
type LCBool = forall a. a -> a -> a

lcTrue :: LCBool
lcTrue = \x _ -> x

lcFalse :: LCBool
lcFalse = \_ y -> y

lcIf :: LCBool -> a -> a -> a
lcIf condition thenCase elseCase = condition thenCase elseCase

-- Step 3: Implement Basic Combinators
i :: a -> a
i = \x -> x

k :: a -> b -> a
k = \x _ -> x

ki :: a -> b -> b
ki = \_ y -> y

-- Step 4: Create Church Numerals
type LCNat = forall a. (a -> a) -> a -> a

lcZero :: LCNat
lcZero = \_ x -> x

lcSucc :: LCNat -> LCNat
lcSucc n = \f x -> f (n f x)

churchToInt :: LCNat -> Int
churchToInt n = n (+1) 0

intToChurch :: Int -> LCNat
intToChurch 0 = lcZero
intToChurch n = lcSucc (intToChurch (n - 1))

-- Step 5: Implement Basic Arithmetic
lcAdd :: LCNat -> LCNat -> LCNat
lcAdd m n = \f x -> m f (n f x)

lcMult :: LCNat -> LCNat -> LCNat
lcMult m n = \f -> m (n f)

-- Step 6: Implement Pairs
type LCPair a b = forall c. (a -> b -> c) -> c

lcPair :: a -> b -> LCPair a b
lcPair x y = \f -> f x y

lcFirst :: LCPair a b -> a
lcFirst p = p (\x _ -> x)

lcSecond :: LCPair a b -> b
lcSecond p = p (\_ y -> y)

-- Step 7: Build Lists Using Pairs
type LCList a = forall b. (a -> b -> b) -> b -> b

lcNil :: LCList a
lcNil = \_ n -> n

lcCons :: a -> LCList a -> LCList a
lcCons x xs = \c n -> c x (xs c n)

lcIsNil :: LCList a -> LCBool
lcIsNil xs = xs (\_ _ -> lcFalse) lcTrue

lcHead :: LCList a -> a
lcHead xs = xs (\x _ -> x) (error "Empty list")

lcTail :: LCList a -> LCList a
lcTail xs = lcSecond (xs (\x r -> lcPair x (lcFirst r)) (lcPair (error "Empty list") lcNil))

-- Step 8: Implement Comparison Operations
lcIsZero :: LCNat -> LCBool
lcIsZero n = n (\_ -> lcFalse) lcTrue

lcLessThan :: LCNat -> LCNat -> LCBool
lcLessThan m n = lcIsZero (lcSubtract m n)

lcEqual :: LCNat -> LCNat -> LCBool
lcEqual m n = lcAnd (lcLessThan m (lcSucc n)) (lcLessThan n (lcSucc m))

-- Helper functions for comparison
lcSubtract :: LCNat -> LCNat -> LCNat
lcSubtract m n = n lcPred m
  where
    lcPred :: LCNat -> LCNat
    lcPred n = lcFirst (n (\p -> lcPair (lcSecond p) (lcSucc (lcSecond p))) (lcPair lcZero lcZero))

lcAnd :: LCBool -> LCBool -> LCBool
lcAnd p q = p q lcFalse

-- Step 9: Implement Higher-Order List Operations
lcMap :: (a -> b) -> LCList a -> LCList b
lcMap f xs = xs (\x r -> lcCons (f x) r) lcNil

lcFilter :: (a -> LCBool) -> LCList a -> LCList a
lcFilter p xs = xs (\x r -> lcIf (p x) (lcCons x r) r) lcNil

lcFold :: (a -> b -> b) -> b -> LCList a -> b
lcFold f z xs = xs f z

-- Step 10: Explore Recursion and Fixed-Point Combinators
fix :: (a -> a) -> a
fix f = let x = f x in x

-- Example: Factorial using fix
lcFactorial :: LCNat -> LCNat
lcFactorial = fix $ \f n ->
  lcIf (lcIsZero n)
       (intToChurch 1)
       (lcMult n (f (lcSubtract n (intToChurch 1))))

-- Helper function to convert LCList of LCNat to regular Haskell list of Int
lcListToList :: LCList LCNat -> [Int]
lcListToList xs = lcFold (\x acc -> churchToInt x : acc) [] xs

-- Example usage
main :: IO ()
main = do
  putStrLn "Church numeral examples:"
  print $ churchToInt (lcSucc lcZero)  -- Should output 1
  print $ churchToInt (lcAdd (intToChurch 2) (intToChurch 3))  -- Should output 5
  print $ churchToInt (lcMult (intToChurch 2) (intToChurch 3))  -- Should output 6

  putStrLn "\nChurch boolean examples:"
  print $ lcIf lcTrue "yes" "no"  -- Should output "yes"
  print $ lcIf lcFalse "yes" "no"  -- Should output "no"

  putStrLn "\nList examples:"
  let list123 = lcCons (intToChurch 1) (lcCons (intToChurch 2) (lcCons (intToChurch 3) lcNil))
  print $ lcListToList list123  -- Should output [1,2,3]
  print $ lcListToList (lcMap (lcAdd (intToChurch 1)) list123)  -- Should output [2,3,4]
  print $ lcListToList (lcFilter (\n -> lcLessThan n (intToChurch 3)) list123)  -- Should output [1,2]

  putStrLn "\nFactorial example:"
  print $ churchToInt (lcFactorial (intToChurch 5))  -- Should output 120